python/tchannel/io.py

```diff
@@ -19,8 +19,169 @@
+    from toro import Condition
+
+from .messages import common
+
+
+class BaseArgStream(object):
+
+    def __init__(self):
+        """Base Arg Stream Class
+
+        _state is used to track the state of the arg stream.
+
+        When object is created, it is in the "init" state.
+
+        When write func is called, it will transformed to "streaming" state.
+
+        When close func is called or reach the end of the stream,
+        it will jump to "completed" state no matter what state it is before.
+
+        If the stream is closed, it won't allow to write anymore.
+        If the stream is closed, it will return whatever still remains in the
+        stream buffer, after that it will just return empty.
+        """
+        self._state = StreamState.init
+
+    def read(self):
+        """Async read from internal stream buffer
+
+        if it reaches the end of the stream, it will return empty(""), caller
+        will depend on the return value to tell if it reaches the end of the
+        stream.
+
+        It doesn't support seek functionality, which means it will only read
+        the stream data once in single direction.
+
+        NOTE: The data which has been read will be discarded in the stream.
+
+        :return:chunk of bytes read from stream
+        """
+        raise NotImplementedError()
+
+    def write(self, chunk):
+        """Async write to internal stream buffer
+
+        :return if stream has been closed, it will raise StreamingException
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        """Close the stream
+
+        It updates the _state to completed.
+
+        :return:
+        """
+        raise NotImplementedError()
+
+
+class InMemArgStream(BaseArgStream):
+
+    def __init__(self, buf=None, auto=True):
+        """In-Memory based arg stream
+
+        :param buf:
+        :return:
+        """
+        super(InMemArgStream, self).__init__()
+        self._stream = buf or ""
+        self._condition = Condition()
+        self.auto = auto
+
+    @tornado.gen.coroutine
+    def read(self):
+        if self._state != StreamState.completed and self._stream == "":
+            yield self._condition.wait()
+
+        chunk = self._stream
+        self._stream = ""
+        raise tornado.gen.Return(chunk)
+
+    @tornado.gen.coroutine
+    def write(self, chunk):
+        if self._state == StreamState.completed:
+            raise StreamingException("Stream has been closed.")
+
+        self._stream += chunk
+        self._condition.notify()
+
+    @tornado.gen.coroutine
+    def close(self):
+        self._state = StreamState.completed
+        self._condition.notify()
+
+
+class PipeArgStream(BaseArgStream):
+
+    def __init__(self, rpipe, wpipe=None, auto=False):
+        """Pipe-based arg stream
+
+        NOTE: if you want to pass a file descriptor of a file, you have to
+        use os.open to get the file descriptor instead of python's open.
```

Comment:
Since this will accept any file descriptor and the problem exists only for `open`-based file descriptors, just mention that for reading from or writing to files, `os.open` must be used instead of `open`. Socket file descriptors and others will be just fine.